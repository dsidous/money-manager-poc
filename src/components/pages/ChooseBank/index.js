import React, { PureComponent } from 'react';
import {
  View,
  TouchableOpacity,
  Image,
  FlatList,
  StyleSheet,
} from 'react-native';
import { connect } from 'react-redux';
import { StyledText, StyledButton } from '../../atoms';
import { FETCH_BANK_DATA, SELECTED_BANK_DATA } from "../../../store/constants";
import bankImages from '../../../assets/images/banklogos/index';
import PropTypes from 'prop-types';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    marginTop: 20,
    paddingHorizontal: 25,
  },
  bankItem: {
    flexDirection: 'row',
    justifyContent: "center",
    alignItems: "center",
    padding: 20,
    marginVertical: 10,
    backgroundColor: '#fff',
    borderColor: '#D3D3D3',
    borderWidth: 1,
    borderRadius: 5
  },

  bankItemSelected: {
    backgroundColor: '#ccebff',
    borderColor: '#00f'
  },

});

class ChooseBank extends PureComponent {

  componentDidMount() {
    this.props.dispatch({ type: FETCH_BANK_DATA })
  }

  renderListItem = ({ item }) => {
    let selected = '';

    if (this.props.selectedBank && this.props.selectedBank.bankName === item.bankName) {
      selected = styles.bankItemSelected;
    }

    return (
      <TouchableOpacity
        onPress={() => this.props.dispatch({ type: SELECTED_BANK_DATA, payload: item })}
        style={[styles.bankItem, selected]}
        data-test="bank-list-item"
      >
        <View style={{ width: 50, height: 50, margin: 5, flex: 1 }}>
          <Image style={{ width: '100%', height: '100%', resizeMode: 'contain' }} source={bankImages[item.bankName]} />
        </View>
        <StyledText style={{ flex: 4 }}>
          {item.bankName}
        </StyledText>
      </TouchableOpacity>
    )
  }
  render() {
    return (
      <View style={styles.container}>
        <StyledText size='large' style={{ marginVertical: 15 }}>
          Choose your bank
        </StyledText>
        <StyledText style={{ marginVertical: 15 }}>
          Select your bank that your day to day spending account is with.
        </StyledText>
        <FlatList
          style={{ marginBottom: 50 }}
          data={this.props.bankList}
          renderItem={this.renderListItem}
          keyExtractor={(item) => item.bankName}
        />
        <View style={{ paddingHorizontal: 80 }}>
          <StyledButton
            disabled={!this.props.selectedBank}
            onPress={
              () => this.props.navigation.navigate('Consent')
            }
            data-test="nextButton"
          >
            Connect to bank
          </StyledButton>
        </View>
      </View >
    );
  }
}

const mapStateToProps = (state) => ({
  bankList: state.bank.list,
  selectedBank: state.bank.selected
})

export default connect(mapStateToProps)(ChooseBank);

ChooseBank.propTypes = {
  navigation: PropTypes.object,
  selectedBank: PropTypes.object,
  bankList: PropTypes.array,
}