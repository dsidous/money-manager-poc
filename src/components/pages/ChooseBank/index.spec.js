import React from 'react';
import { shallow, configure } from 'enzyme';
import configureStore from 'redux-mock-store';
import Adapter from 'enzyme-adapter-react-16';
import TestRenderer from 'react-test-renderer';
import ChooseBank from '.';

configure({ adapter: new Adapter() });

describe('choose bank', () => {
  const initialState = {
    bank: {
      list: [
        { bankName: 'bank1', logoImage: 'bank1logo' },
        { bankName: 'bank2', logoImage: 'bank2logo' },
        { bankName: 'bank3', logoImage: 'bank3logo' },
      ],
      selected: {
        bankName: 'bank1',
        logoImage: 'bank1logo',
      },
    },
  };
  const mockStore = configureStore();
  const store = mockStore(initialState);
  const wrapper = shallow(<ChooseBank store={store} />);


  it('renders as expected', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Selecting first item and navigate to next screen', () => {
    const navigation = { navigate: jest.fn() };
    const testrenderer = TestRenderer.create(<ChooseBank navigation={navigation} store={store} />);
    const testinstance = testrenderer.root;

    testinstance.findAllByProps({ 'data-test': 'bank-list-item' })[0].props.onPress();
    expect(testinstance.findByProps({ 'data-test': 'nextButton' }).props.disabled).toBe(false);

    testinstance.findByProps({ 'data-test': 'nextButton' }).props.onPress();
    expect(navigation.navigate).toHaveBeenCalledWith('Consent');
  });
});
