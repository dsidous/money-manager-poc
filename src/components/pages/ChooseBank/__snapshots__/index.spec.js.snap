// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`choose bank renders as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(ChooseBank)
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "bankList": Array [
        Object {
          "bankName": "bank1",
          "logoImage": "bank1logo",
        },
        Object {
          "bankName": "bank2",
          "logoImage": "bank2logo",
        },
        Object {
          "bankName": "bank3",
          "logoImage": "bank3logo",
        },
      ],
      "dispatch": [Function],
      "selectedBank": Object {
        "bankName": "bank1",
        "logoImage": "bank1logo",
      },
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeSubscription": Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "bankList": Array [
          Object {
            "bankName": "bank1",
            "logoImage": "bank1logo",
          },
          Object {
            "bankName": "bank2",
            "logoImage": "bank2logo",
          },
          Object {
            "bankName": "bank3",
            "logoImage": "bank3logo",
          },
        ],
        "dispatch": [Function],
        "selectedBank": Object {
          "bankName": "bank1",
          "logoImage": "bank1logo",
        },
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "storeSubscription": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
